internal/threadgroup/threadgroup.go

```diff
@@ -0,0 +1,96 @@
+package threadgroup
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+type (
+	// A ThreadGroup provides synchronization between a module and its
+	// goroutines to enable clean shutdowns
+	ThreadGroup struct {
+		mu     sync.Mutex
+		wg     sync.WaitGroup
+		closed chan struct{}
+	}
+)
+
+// ErrClosed is returned when the threadgroup has already been stopped
+var ErrClosed = errors.New("threadgroup closed")
+
+// Done returns a channel that will be closed when the threadgroup is stopped
+func (tg *ThreadGroup) Done() <-chan struct{} {
+	return tg.closed
+}
+
+// Add adds a new thread to the group, done must be called to signal that the
+// thread is done. Returns ErrClosed if the threadgroup is already closed.
+func (tg *ThreadGroup) Add() (func(), error) {
+	tg.mu.Lock()
+	defer tg.mu.Unlock()
+	select {
+	case <-tg.closed:
+		return nil, ErrClosed
+	default:
+	}
+	tg.wg.Add(1)
+	return func() { tg.wg.Done() }, nil
+}
+
+// WithContext returns a copy of the parent context. The returned context will
+// be cancelled if the parent context is cancelled or if the threadgroup is
+// stopped.
+func (tg *ThreadGroup) WithContext(parent context.Context) (context.Context, context.CancelFunc) {
+	// wrap the parent context in a cancellable context
+	ctx, cancel := context.WithCancel(parent)
```

Comment:
I may be wrong, but I didn't think you could check `ctx.Err()` for the error message returned by cancel cause.

Reply:
Wow TIL. Definitely got a few spots in `renterd` to update now. Returning `context.Cause(ctx)` instead of `ctx.Err` isn't too bad though. Especially since `context.Cause` will return the same error for cancellation without custom error as `ctx.Err`.